Add debug feature module section + require inspector (#1089)
spatialos/gdk-for-unity: docs/modules/debug/require-inspector.md

<%( TOC )%>

> # Require inspector extension

Reviewer: I'm not sure about this name. Need to think about it a bit

Author: removed the extension bit here and in the TOC

> The Require inspector extension is a debugging extension for the Unity inspector, which helps with visualizing the use of `[Require]` in MonoBehaviour components.

Reviewer: this needs links to the requireables docs 

inspector needs to be captialized according to Unity docs (everywhere you use it in this doc)

## Usage

To use the extension, ensure the package is added to your project and enter Play mode.

> 1. Select any SpatialOS Gameobject in your Hierarchy.

Reviewer: ```suggestion
1. Select any Unity GameObject that is linked to a SpatialOS entity in your Hierarchy.
```

Reviewer: should have a mention somewhere that this only works together with the GameObject module and some docs to there, if they want to find to out more

Author: It works with the subscription system, which is core?

Reviewer: I meant that you can only use it when you have a gameobject linked to a SpatialOS entity -> need to use the gameobject module, wihch is at the moment core.
but maybe adding a link to this point is enough

Reviewer 2: The linking is actually part of the Core right now - https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/io.improbable.gdk.core/Subscriptions/EntityGameObjectLinker.cs 😉

Reviewer 2: This actually raises an interesting question for the future though. When gameobject linking moves out of the Core what does the Debug module depend on? 

Do we have the tooling inside each feature module or bunch up tooling into these Debug feature modules?

> 1. Open the Inspector.

Reviewer: ```suggestion
1. Open the Unity Inspector.
```

1. Browse to a MonoBehaviour that you want to inspect.
1. Select the ▶ button next to `SpatialOS`

<img src="{{assetRoot}}assets/modules/debug/inspector.png" style="margin: 0 auto; width: auto; display: block;" />

## Explanation

> The list of types displayed in the inspector, are the types that the MonoBehaviour requires through its `[Require]` attributes. All of these requirements have to be available, before the MonoBehaviour will be enabled.

Reviewer: objects not types right?  it's a bit weird to require a type

```suggestion
The extension lists all objects that are required through the `[Require]` attributes. The dot in front of each type indicates whether an object is available:
*  Green: The object is available.
*  Grey: The object is not available.

Each object that you require may have requirements that need to be fulfilled, before the GDK is able to make the object available and enable this MonoBehaviour.
```

> The dot in front of each type indicates the current state of this requirement. Green means it is available, and grey means it is not.

Reviewer: ```suggestion
```

If you have a grey indicator on a `Reader` type it can be due to one the following reasons:

> * The entity does not have the Component present.

Reviewer: the term SpatialOS component is not capitalized
```suggestion
* The entity does not have the component present.
```

Reviewer: needs links to entity and component

> * The Worker does not have read access on the component.

Reviewer: ```suggestion
* The worker does not have read access on the component.
```

Reviewer: needs link for worker and read access

> A grey indicator for a `Writer` type is either due to the above reasons, or due to the worker not having authority over the component at that moment.

Reviewer: make this also a bullet point list

Reviewer: we never call it `Writer` type anywhere else. just call it writer and link to the correct docs

<%( TOC )%>

> # Require Inspector

Reviewer 3: is this the name we settled on? 

`Require Inspector` vs `Requirable Inspector`

> The Require Inspector extension is a debugging extension for the Unity Inspector, which helps with visualizing the use of [`[Require]`]({{urlRoot}}/reference/workflows/monobehaviour/interaction/reader-writers/lifecycle) in MonoBehaviour components.

Reviewer 2: ```suggestion
The Require Inspector extension is a debugging extension for the Unity Inspector, which helps visualize the use of [`[Require]`]({{urlRoot}}/reference/workflows/monobehaviour/interaction/reader-writers/lifecycle) in MonoBehaviour components.
```

1. Select the ▶ button next to `SpatialOS`

<img src="{{assetRoot}}assets/modules/debug/inspector.png" style="margin: 0 auto; width: auto; display: block;" />

## Explanation

The extension lists all objects that are required through the [`[Require]`]({{urlRoot}}/reference/workflows/monobehaviour/interaction/reader-writers/lifecycle) attributes. The dot in front of each type indicates whether an object is available:

* Green: The object is available.
* Grey: The object is not available.

Each object that you require may have requirements that need to be fulfilled, before the GDK is able to make the object available and enable this MonoBehaviour.

If you have a grey indicator on a [`Reader`]({{urlRoot}}/reference/workflows/monobehaviour/interaction/reader-writers/index) it can be due to one the following reasons:

* The [entity]({{urlRoot}}/reference/glossary#spatialos-entity) does not have the [component]({{urlRoot}}/reference/glossary#spatialos-component) present.
* The [worker instance]({{urlRoot}}/reference/glossary#worker) does not have [read access]({{urlRoot}}/reference/glossary#read-access) on the [component]({{urlRoot}}/reference/glossary#spatialos-component).

> A grey indicator for a [`Writer`]({{urlRoot}}/reference/workflows/monobehaviour/interaction/reader-writers/index) can be due to:

Reviewer 2: Add these for the `CommandReceiver` too? It basically has the same requirements as a writer
